sklearn_numba_dpex/common/kernels.py

```diff
@@ -760,13 +738,19 @@ def partial_sum_reduction(
         # a similar memory access pattern than seen at the previous step.
         n_active_sub_groups = n_sub_groups_per_work_group
         for i in range(n_local_iterations):
+
             # At each iteration, half of the remaining work items with the highest id
             # are discarded.
             n_active_sub_groups = n_active_sub_groups // two_as_a_long
             work_item_row_idx = first_row_idx + local_row_idx + n_active_sub_groups
             if (
                 (local_row_idx < n_active_sub_groups) and
-                (col_idx < n_cols) and
+                # HACK: this condition save some memory access and was added thinking
+                # it might be good for performance. However on CPU it causes wrong
+                # results in some tests.
+                # TODO: create a minimal reproducer and report to `numba_dpex`. Remove
+                # the hack once it is fixed.
+                (_is_cpu or (col_idx < n_cols)) and
```

Comment:
There's probably some bug in numba_dpex or with the cpu backend here that causes branching to not be correctly executed in very specific cases, for this very particular condition, on cpu. I had no luck when trying to have a minimal reproducer and I can't see failures on gpu. So I'd be in favor of letting this pass through in the meantime. Note that this also fixes the case with `work_group_size=1` that I added back to the unit tests.

Reply:
Well it's still wrong on the CI. :shrug: